api/swagger.yml

```diff
@@ -1986,6 +2020,76 @@ paths:
           schema:
             $ref: "#/definitions/error"
 
+  /repositories/{repository}/branches/{branch}/staging/backing:
+    parameters:
+      - in: path
+        name: repository
+        required: true
+        type: string
+      - in: path
+        name: branch
+        required: true
+        type: string
+      - in: query
+        name: path
+        required: true
+        type: string
+    get:
+      tags:
+        - staging
+      operationId: getPhysicalAddress
+      summary: get a physical address and a return token to write object to underlying storage
+      responses:
+        200:
+          description: physical address for staging area
+          schema:
+            $ref: "#/definitions/staging_location"
+        401:
+          $ref: "#/responses/Unauthorized"
+        404:
+          description: branch not found
+          schema:
+            $ref: "#/definitions/error"
+        default:
+          description: generic error response
+          schema:
+            $ref: "#/definitions/error"
+    put:
+      tags: 
+       - staging
+      operationId: linkPhysicalAddress
+      summary: associate staging on this physical address with a path
+      description: |
+        If the supplied token matches the current staging token, associate the object as the
+        physical address with the supplied path.  Otherwise return a conflict and hint where
+        to place the object to try again.
+      parameters:
+        - in: body
+          name: metadata
+          schema:
+            $ref: "#/definitions/staging_metadata"
+      responses:
+        200:
+          # This actually violates HTTP, which requires returning 201 if a new object was
+          # created or 200 if an existing object was modified,
+          # https://tools.ietf.org/html/rfc7231#section-4.3.4
+          description: successfully linked
+        401:
+          $ref: "#/responses/Unauthorized"
+        404:
+          description: branch not found
+          schema:
+            $ref: "#/definitions/error"
+        409:
+          description: conflict with a commit, try here
+          schema:
+            $ref: "#/definitions/staging_location"
```

Comment:
This implies there's a certain flow of requests that should take place between client and server. The happy path is quite clear, but assuming I'm writing a client, how should I handle this response? What should I do with the data I've already uploaded? Should I clean that up? will the lakeFS server do it for me? How often should I retry? should I employ some sort of backoff? At the very least we should document these.

Reply:
Sure, will document.
In brief:
- It says "try here", you should try there.  Or give up -- I'm an API response, not a cop.
- Cleanup is deferred to when we decide where to upload, and how to handle lifecycle in general.  I will _recommend_ that clients copy the file to the recommended location (should be fast if intra-region), delete the original, and call again.
- The lakeFS won't do anything for you if it asks you to try again, so no.  In general I don't think the server should do data operations for the client if the client is in "let me do my own data operations" mode.
- There is no need to backoff on retries: this will not be an idempotent retry, and it's not due to load.  The server is asking the client to try again with _different_ parameters.